prov/efa/src/efa_cq.c

```diff
@@ -121,65 +139,57 @@ static void efa_cq_read_data_entry(struct efa_wc *wc, int i, void *buf)
 ssize_t efa_cq_readfrom(struct fid_cq *cq_fid, void *buf, size_t count,
 			fi_addr_t *src_addr)
 {
+	bool should_end_poll = false;
 	struct efa_cq *cq;
-	struct efa_wce *wce;
-	struct slist_entry *entry;
 	struct efa_av *av;
 	struct efa_wc wc = {0};
-	ssize_t ret = 0, i;
+	ssize_t err = 0;
+	size_t num_cqe = 0; /* Count of read entries */
+
+	/* Initialize an empty ibv_poll_cq_attr struct for ibv_start_poll.
+	 * EFA expects .comp_mask = 0, or otherwise returns EINVAL.
+	 */
+	struct ibv_poll_cq_attr poll_cq_attr = {.comp_mask = 0};
 
 	cq = container_of(cq_fid, struct efa_cq, util_cq.cq_fid);
 
 	ofi_spin_lock(&cq->lock);
 
-	for (i = 0; i < count; i++) {
-		if (!slist_empty(&cq->wcq)) {
-			wce = container_of(cq->wcq.head, struct efa_wce, entry);
-			if (wce->wc.ibv_wc.status) {
-				ret = -FI_EAVAIL;
-				break;
-			}
-			entry = slist_remove_head(&cq->wcq);
-			wce = container_of(entry, struct efa_wce, entry);
-			cq->read_entry(&wce->wc, i, buf);
-			ofi_buf_free(wce);
-			continue;
-		}
+	/* Call ibv_start_poll only once regardless of count == 0 */
+	err = ibv_start_poll(cq->ibv_cq_ex, &poll_cq_attr);
+	should_end_poll = !err;
 
-		ret = ibv_poll_cq(cq->ibv_cq, 1, &wc.ibv_wc);
-		if (ret != 1) {
-			if (!ret)
-				ret = -FI_EAGAIN;
-			break;
-		}
+	while (!err && num_cqe < count) {
+		efa_cq_wc_from_ibv_cq_ex_unsafe(cq->ibv_cq_ex, &wc.ibv_wc);
 
-		/* Insert error entry into wcq */
-		if (wc.ibv_wc.status) {
-			wce = ofi_buf_alloc(cq->wce_pool);
-			if (!wce) {
-				ofi_spin_unlock(&cq->lock);
-				return -FI_ENOMEM;
-			}
-			memset(wce, 0, sizeof(*wce));
-			memcpy(&wce->wc, &wc, sizeof(wc));
-			slist_insert_tail(&wce->entry, &cq->wcq);
-			ret = -FI_EAVAIL;
+		if (cq->ibv_cq_ex->status) {
+			err = -FI_EAVAIL;
 			break;
 		}
 
 		if (src_addr) {
 			av = cq->domain->qp_table[wc.ibv_wc.qp_num &
-			     cq->domain->qp_table_sz_m1]->ep->av;
+				cq->domain->qp_table_sz_m1]->ep->av;
 
-			src_addr[i] = efa_av_reverse_lookup_dgram(av,
-								  wc.ibv_wc.slid,
-								  wc.ibv_wc.src_qp);
+			src_addr[num_cqe] = efa_av_reverse_lookup_dgram(av,
+				wc.ibv_wc.slid,
+				wc.ibv_wc.src_qp);
 		}
-		cq->read_entry(&wc, i, buf);
+
+		cq->read_entry(&wc, num_cqe, buf);
+		num_cqe++;
+
+		err = ibv_next_poll(cq->ibv_cq_ex);
 	}
 
+	err = (err == ENOENT) ? -FI_EAGAIN : (err > 0 ? -err : err);
```

Comment:
This is too complicated for 1 line, can you wrap it into an inline function?

Reply:
Sure. Will address.